From 736d91bef1633c9b1010675d40f1f8f9ba0e4333 Mon Sep 17 00:00:00 2001
From: Christophe Roullier <christophe.roullier@foss.st.com>
Date: Thu, 19 Dec 2024 15:50:35 +0100
Subject: [PATCH] net: phy: call phy_start always in good state.

Avoid to execute phy_start when PHY already in Running state.

Change-Id: I33b9d14bd2fa5e53c50f6cc78540de4f81a4dda4
Signed-off-by: Christophe Roullier <christophe.roullier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/428097
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 drivers/net/phy/phy.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/net/phy/phy.c b/drivers/net/phy/phy.c
index df54c137c5f5..9d14cb87518e 100644
--- a/drivers/net/phy/phy.c
+++ b/drivers/net/phy/phy.c
@@ -1411,7 +1411,10 @@ void phy_start(struct phy_device *phydev)
 {
 	mutex_lock(&phydev->lock);
 
-	if (phydev->state != PHY_READY && phydev->state != PHY_HALTED) {
+	if (phydev->state == PHY_RUNNING || phydev->state == PHY_UP)
+		goto out;
+
+	if (phydev->state != PHY_READY && phydev->state != PHY_HALTED && phydev->state != PHY_NOLINK) {
 		WARN(1, "called from state %s\n",
 		     phy_state_to_str(phydev->state));
 		goto out;
-- 
2.50.1

